qiskit/extensions/standard/x.py

```diff
@@ -414,3 +447,455 @@ def ccx(self, control_qubit1, control_qubit2, target_qubit,
 # support both ccx and toffoli as methods of QuantumCircuit
 QuantumCircuit.ccx = ccx
 QuantumCircuit.toffoli = ccx
+
+
+class CCCXGate(ControlledGate):
+    """The 3-qubit controlled X gate.
+
+    This implementation is based on Page 17 of [1].
+
+    References:
+        [1] Barenco et al., 1995. https://arxiv.org/pdf/quant-ph/9503016.pdf
+    """
+
+    def __init__(self, angle=numpy.pi/4):
+        """Create a new 3-qubit controlled X gate.
+
+        Args:
+            angle (float): The angle used in the controlled-U1 gates. An angle of π/4 yields the
+                3-qubit controlled X gate, an angle of π/8 the 3-qubit controlled sqrt(X) gate.
+        """
+        super().__init__('cccx', 4, [], num_ctrl_qubits=3)
+        self.base_gate = XGate()
+        self._angle = angle
+
+    def _define(self):
+        """
+        gate cccx a,b,c,d
+        {
+            h d; cu1(-pi/4) a,d; h d;
+            cx a,b;
+            h d; cu1(pi/4) b,d; h d;
+            cx a,b;
+            h d; cu1(-pi/4) b,d; h d;
+            cx b,c;
+            h d; cu1(pi/4) c,d; h d;
+            cx a,c;
+            h d; cu1(-pi/4) c,d; h d;
+            cx b,c;
+            h d; cu1(pi/4) c,d; h d;
+            cx a,c;
+            h d; cu1(-pi/4) c,d; h d;
+        }
+
+        gate cccsqrtx a,b,c,d
+        {
+            h d; cu1(-pi/8) a,d; h d;
+            cx a,b;
+            h d; cu1(pi/8) b,d; h d;
+            cx a,b;
+            h d; cu1(-pi/8) b,d; h d;
+            cx b,c;
+            h d; cu1(pi/8) c,d; h d;
+            cx a,c;
+            h d; cu1(-pi/8) c,d; h d;
+            cx b,c;
+            h d; cu1(pi/8) c,d; h d;
+            cx a,c;
+            h d; cu1(-pi/8) c,d; h d;
+        }
```

Comment:
this block seems like it belongs to CCCCX further down

Reply:
Yes, I'll move it there